master/internal/api_experiment.go

```diff
@@ -1922,6 +1934,477 @@ func sortExperiments(sortString *string, experimentQuery *bun.SelectQuery) error
+	valueTypeKnown := false
+	filterIndex := int32(startIndex)
+	valueHasStarted := valueStart
+	isCol := !valueHasStarted
+	isNull := false
+
+	for {
+		if filterIndex > int32(len(filter)-1) {
+			break
+		}
+		c := string(filter[filterIndex])
+
+		if valueHasStarted && valueIsString && valueTypeKnown {
+			// If the current value is a string and we have reached
+			// a string terminating character then we are at the end
+			// of the value
+			if c == "'" || c == "\"" {
+				filterIndex++
+				break
+			}
+		} else if valueHasStarted && !valueIsString && valueTypeKnown {
+			// If the current value is a number and the current character
+			// is not numeric then we are at the end of the value
+			if !numericeRegex.MatchString(c) {
+				break
+			}
+		}
+		if !valueTypeKnown && valueHasStarted {
+			// The start of the value has been reached so the type of the value
+			// can be determined
+			valueTypeKnown = true
+			if c == "'" || c == "\"" { //nolint: gocritic
+				// If the value starts with a quote character then the value is a string
+				valueIsString = true
+				filterIndex++
+				continue
+			} else if filterIndex+3 < int32(len(filter)) &&
+				filter[filterIndex:filterIndex+4] == metricTypeNull {
+				// If the value from this point forward is 'null' then the value is 'NULL'
+				filterIndex += 4
+				isNull = true
+				break
+			} else {
+				// Otherwise the value is a number
+				valueIsString = false
+			}
+		}
+		if c == ":" || c == "~" {
+			// These characters represent the end of a column name
+			// and the start of the value
+			if isCol {
+				isCol = false
+				valueHasStarted = true
+				filterIndex++
+				comparator = c
+				continue
+			}
+		}
+		if c == "\\" {
+			if filterIndex+1 < int32(len(filter)-1) {
+				nextChar := string(filter[filterIndex+1])
+				// Check to see if a terminated quote has been reached
+				// if so then add it to the sql string.
+				if nextChar == `'` || nextChar == `"` {
+					c = `\` + nextChar
+				}
+				filterIndex++
+			}
+		}
+		switch isCol {
+		case true:
+			col += c
+		case false:
+			value += c
+		}
+		filterIndex++
+	}
+	if *operator == ":" {
+		if isNull {
+			query = " IS NULL"
+		} else {
+			if valueIsString {
+				value = fmt.Sprintf(`'%s'`, value)
+			}
+			query = fmt.Sprintf(" = %v", value)
+		}
+	}
+	if *operator == "~" {
+		query = " LIKE " + `'%` + value + `%'`
+	}
+	if *operator == "-" {
+		if isNull { //nolint: gocritic
+			query = fmt.Sprintf("%v IS NOT NULL", col)
+		} else if comparator == "~" {
+			query = fmt.Sprintf("%v NOT LIKE ", col) + `'%` + value + `%'`
+		} else if comparator == ":" {
+			if valueIsString {
+				value = `'` + value + `'`
+			}
+			query = fmt.Sprintf("%v != %v", col, value)
+		}
+	}
+	if *operator == "<" || *operator == ">" {
+		if valueIsString {
+			value = fmt.Sprintf(`'%s'`, value)
+		}
+		query = value
+	}
+	return query, filterIndex
+}
+
+func parseMetricComparisons(filter string) (map[string]string, error) {
```

Comment:
Ideally we should just accept some structured JSON here but if you're going to write a parser you should at least be using a [parser library](https://pkg.go.dev/search?q=parser+combinator).